### PR TITLE
Fix getCurrentUrl() not respecting store base path

### PR DESCRIPTION
**Pull request by toxix**
This is correcting an issue for magento where magentos
\Magento\Framework\Url::getCurrentUrl()
function is not respecting the stores base url.

### Background
$this->_request->getOriginalPathInfo();
In setups, where the stores are
switched with the setting of
  SetEnv MAGE_RUN_TYPE store
  SetEnv MAGE_RUN_CODE de_de
the path that is available in the php request object is in our setup
not including the complete path but only the path after the
storeBaseUrl.

 url in browser: https://example.com/de/shop/hosen/
 host in php: example.com
 path in php: /hosen/
 base url in Magento: https://example.com/de/shop/

Now we running into an issue with the add to compare list links. This
links are generated and the redirect url is set already in the link
itself with the 'uenc' parameter. And this parameter contains a wrong
url: https://example.com/hosen/

After tracing down where this wrong url is comming from, it looks like
in the getCurrentUrl function of magento the base url of the shop is not
respected. Therefor I decided to change this function in a way that is
respecting the store base url if it is set and falls back to server/php
url otherwise.

### additional changes
Additionaly this also refector the url generation in a way, that is not
relying on manualy determine the parts of the url, but using the
zend/php framework for this.

### Feedback
This Fix is not tested with a Multistore setup that is based on Symlinks. I am not 100% sure if the 
`$this->_request->getOriginalPathInfo()` is delivering the correct path in this setups.

Not sure on how to run the automated tests.

I have also checked the usages of this function and would not expect any mayor side effects.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=23854) <br/>All committers have signed the CLA.

**Comment by josefbehr** [comment]
@toxix is there a confirmed bug related to this PR? If not, can you please create one and post steps to reproduce the issue?

File: app/code/Magento/Backend/Model/Url.php [hunk 1]
```diff
@@ -126,6 +127,7 @@ public function __construct(
         \Magento\Backend\Model\Auth\Session $authSession,
         \Magento\Framework\Encryption\EncryptorInterface $encryptor,
         \Magento\Store\Model\StoreFactory $storeFactory,
+        \Magento\Store\Model\StoreManagerInterface $storeManager,
```
Comment (sidolov) [inline app/code/Magento/Backend/Model/Url.php:130]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

File: lib/internal/Magento/Framework/Url.php [hunk 2]
```diff
@@ -93,6 +93,11 @@ class Url extends \Magento\Framework\DataObject implements \Magento\Framework\Ur
         '_scope_to_url',
     ];
 
+    /**
+     * @var \Magento\Store\Model\StoreManagerInterface
+     */
+    protected $_storeManager;
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Url.php:99]: Please, avoid usage underscore in the variables name and make the property private

File: lib/internal/Magento/Framework/Url.php [hunk 3]
```diff
@@ -221,6 +227,7 @@ public function __construct(
         \Magento\Framework\Url\QueryParamsResolverInterface $queryParamsResolver,
         \Magento\Framework\App\Config\ScopeConfigInterface $scopeConfig,
         \Magento\Framework\Url\RouteParamsPreprocessorInterface $routeParamsPreprocessor,
+        \Magento\Store\Model\StoreManagerInterface $storeManager,
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Url.php:230]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by sidolov** [comment]
@toxix I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**